### DISCUSSION TITLE
Is it okey that CodeBlockAction is called for ArrowExpressionClause? (PropertyDeclaration is expected)

### DISCUSSION THREAD

**Discussion by Denis535** [dotnet/roslyn#49898] [Q&A]
I've made a simple analyzer.

**UnitTest:**

    public class ExampleAnalyzerTests {

       private const string FakeSourceCode = @"
    using System;
    using System.Collections.Generic;
    using System.Text;

    namespace Namespace {
        public class Class {
            public object Field = null;
            public object Prop {get;set;} = null;
            public object Prop2 => null; // Here is strange behavior
            public event Action Event = null;
            public void Method() {
                static void LocalMethod() {}
            }
        }
    }";

        private Solution Solution { get; set; }

        [SetUp]
        public void SetUp() {
            Trace.Listeners.Add( new TextWriterTraceListener( TestContext.Out ) );
            Solution = CreateFakeSolution();
        }
        [TearDown]
        public void TearDown() {
            Solution = null;
        }

        [Test]
        public async Task Test_00() {
            var analyzers = ImmutableArray.Create<DiagnosticAnalyzer>( new ExampleAnalyzer1() );
            var diagnostics = await Analyze( Solution.Projects.Single(), analyzers ).ConfigureAwait( false );
            Assert.That( diagnostics, Is.Empty );
        }

        // Helpers
        private static Solution CreateFakeSolution() {
            return new AdhocWorkspace()
                .AddSolution( SolutionInfo.Create( SolutionId.CreateNewId(), VersionStamp.Create(), null, null, null ) )
                .AddProject( "FakeProject", "FakeProject", LanguageNames.CSharp )
                .AddMetadataReference( MetadataReference.CreateFromFile( typeof( object ).Assembly.Location ) )
                .AddDocument( "FakeClass.cs", FakeSourceCode ).Project.Solution;
        }
        private static async Task<Diagnostic[]> Analyze(Project project, ImmutableArray<DiagnosticAnalyzer> analyzers, CancellationToken cancellationToken = default) {
            var compilation = await project.GetCompilationAsync( cancellationToken ).ConfigureAwait( false );
            var compilationWithAnalyzers = compilation.WithAnalyzers( analyzers, project.AnalyzerOptions );
            var diagnostics = await compilationWithAnalyzers.GetAllDiagnosticsAsync( cancellationToken ).ConfigureAwait( false );
            return diagnostics.Where( i => !IsCompilerDiagnostic( i ) ).ToArray();
        }
        private static bool IsCompilerDiagnostic(Diagnostic diagnostic) {
            return diagnostic.Descriptor.CustomTags.Contains( WellKnownDiagnosticTags.Compiler );
        }

    }

**Analyzer:**

    [DiagnosticAnalyzer( LanguageNames.CSharp )]
    public class ExampleAnalyzer1 : DiagnosticAnalyzer {

        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
            "ExampleAnalyzer1",
            "Example analyzer",
            "Message: {0}",
            "Example",
            DiagnosticSeverity.Warning,
            true );

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create( Rule );

        public override void Initialize(AnalysisContext context) {
            context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.None );
            //context.EnableConcurrentExecution();

            Trace.WriteLine( "DiagnosticAnalyzer" );
            // Compilation
            context.RegisterCompilationStartAction( OnCompilation_Start );
            context.RegisterCompilationAction( OnCompilation );
            // SyntaxTree, SemanticModel
            context.RegisterSyntaxTreeAction( OnSyntaxTree );
            context.RegisterSemanticModelAction( OnSemanticModel );
            // Symbol
            context.RegisterSymbolStartAction( OnSymbol_Start, SymbolKind.NamedType );
            context.RegisterSymbolAction( OnSymbol, (SymbolKind[]) Enum.GetValues( typeof( SymbolKind ) ) );
            // CodeBlock
            context.RegisterCodeBlockStartAction<SyntaxKind>( OnCodeBlock_Start );
            context.RegisterCodeBlockAction( OnCodeBlock );
            // SyntaxNode
            context.RegisterSyntaxNodeAction( OnSyntaxNode, (SyntaxKind[]) Enum.GetValues( typeof( SyntaxKind ) ) );
        }

        // Compilation
        private static void OnCompilation_Start(CompilationStartAnalysisContext context) {
            context.RegisterCompilationEndAction( OnCompilation_End );
        }
        private static void OnCompilation(CompilationAnalysisContext context) {
            //Trace.WriteLine( "OnCompilation: " + context.Compilation.AssemblyName );
        }
        private static void OnCompilation_End(CompilationAnalysisContext context) {
        }

        // SyntaxTree
        private static void OnSyntaxTree(SyntaxTreeAnalysisContext context) {
            //Trace.WriteLine( "OnSyntaxTree: " + Path.GetFileName( context.Tree.FilePath ) );
        }
        // SemanticModel
        private static void OnSemanticModel(SemanticModelAnalysisContext context) {
            //Trace.WriteLine( "OnSemanticModel: " + Path.GetFileName( context.SemanticModel.SyntaxTree.FilePath ) );
        }

        // Symbol
        private static void OnSymbol_Start(SymbolStartAnalysisContext context) { // for specific SymbolKind
            context.RegisterSymbolEndAction( OnSymbol_End );
        }
        private static void OnSymbol(SymbolAnalysisContext context) { // for specific SymbolKind list
            //Trace.WriteLine( "OnSymbol: " + context.Symbol.Name );
        }
        private static void OnSymbol_End(SymbolAnalysisContext context) { // for specific SymbolKind
        }

        // CodeBlock
        private static void OnCodeBlock_Start(CodeBlockStartAnalysisContext<SyntaxKind> context) {
            context.RegisterCodeBlockEndAction( OnCodeBlock_End );
        }
        private static void OnCodeBlock(CodeBlockAnalysisContext context) {
            Trace.WriteLine( "OnCodeBlock: " + context.CodeBlock.Kind() );
            Trace.WriteLine( context.CodeBlock.ToString().Indent( 8 ) );
            Trace.WriteLine( "" );
        }
        private static void OnCodeBlock_End(CodeBlockAnalysisContext context) {
        }

        // SyntaxNode
        private static void OnSyntaxNode(SyntaxNodeAnalysisContext context) { // for specific SyntaxKind list
            //Trace.WriteLine( "OnSyntaxNode: " + context.Node.Kind() );
            //Trace.WriteLine( context.Node.ToString().Indent( 8 ) );
            //Trace.WriteLine( "" );
        }

    }

**Logs:**

```
OnCodeBlock: FieldDeclaration
Syntax: public object Field = null;

OnCodeBlock: EventFieldDeclaration
Syntax: public event Action Event = null;

OnCodeBlock: PropertyDeclaration
Syntax: public object Prop {get;set;} = null;

OnCodeBlock: ArrowExpressionClause // Here is strange behavior
Syntax: => null

OnCodeBlock: MethodDeclaration
Syntax:
public void Method() {
    static void LocalMethod() {}
}
```

As you can see, usually code blocks are member declarations.
But for `public object Prop2 => null;` code block is `ArrowExpressionClause`.
It's very strange! Is it by design?

**Comment by huoyaoyuan** [comment]
It looks strange. What happens for `Prop { get => null; }`?
